[Merged by Bors] - feat(linear_algebra/nonsingular_inverse): state Cramer's rule explicitly

ocfnash opened the pull request:
Mostly so that we can add an entry to the Freek 100.


---
Note that we state this using the slightly low-level `matrix.mul_vec` instead of building on top of `matrix.to_lin'`. This is partly because the current status of the import hierarchy (`matrix.to_lin'` is defined in a file importing `nonsingular_inverse.lean`) and partly because the statement in terms of `to_lin'` turns out to need additional ascriptions which, IMHO, render the statement of `cramers_rule` even less readable to the casual reader arriving from the Freek list.

Notwithstanding the above, I am more than happy to be persuaded to do things differently!

Commit:
feat(linear_algebra/nonsingular_inverse): state Cramer's rule explicitly

Mostly so that we can add an entry to the Freek 100.

jcommelin commented on docs/100.yaml:
```diff
@@ -334,6 +334,8 @@
 97:
   title  : Cramer’s Rule
+  decl   : matrix.cramers_rule
+  author : Tim Baanen
```
```suggestion
  author : Anne Baanen
```

jcommelin committed:
Update docs/100.yaml

jcommelin committed:
the elaborator is now smarter

Commit:
Place emphasis on `cramer` in statement of Cramer's rule.

I feel this is cleaner, though it's obviously a subjective call.

Commit:
Drop redundant redeclaration of parameter

Commit:
Fix statement in comment above `cramer`.

ocfnash commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -57,14 +54,13 @@ section cramer
+  If `A ⬝ x = b` has a unique solution in `x`, `cramer_map Aᵀ` sends the vector `b` to `A.det • x`.
   Otherwise, the outcome of `cramer_map` is well-defined but not necessarily useful.
 -/
 def cramer_map (i : n) : α := (A.update_row i b).det
```
I now see that though the comments refer to replacing a column, the implementation actually replaces a row. This is the reason we have a transpose in the statement of Cramer's rule.

I think this should be fixed but I have run out of minutes. I'll take this up again this evening if it is still pending.

Vierkantor replied:
Ah yes, my old nemesis: rows versus columns. I'll push a commit that fixes it.

jcommelin replied:
@Vierkantor Didn't Oliver mean that we should update the code instead of the docstrings?

Vierkantor replied:
Ah of course. I was thinking that the issue was the comments didn't get updated in #3393.

Vierkantor replied:
I tried modifying `cramer_map`, but it causes too much breakage that I can't fix it right now :(

ocfnash replied:
@Vierkantor Thank you very much for attempting the modification! In the end, I didn't have time this evening but I definitely will tomorrow so I'll see if I can resolve everything then.

ocfnash replied:
I have added a commit which redefines `cramer_map` using `update_column` instead of `update_row` and allows me to state `cramers_rule` without having to insert a transpose, as:
```lean
  cramer A b = A.det • A⁻¹.mul_vec b :=
```

So I think this is fine now.

Vierkantor committed:
Fix docstrings to refer to "row" instead of "column"

Commit:
Redefine `cramer` to be based on `update_column` instead of `update_row`.

This really just moves a transpose operation from the definition of
`cramer_map` to the definition of `adjugate`. I claim the new definitions
have the slight advantage of being more conventional.

In the course of doing this I noticed that there still seem to be a few
places where the words 'row' and 'column' are backwards. I fixed the
instances that I encountered but did not attempt an exhaustive search.

jcommelin commented on src/linear_algebra/determinant.lean:
```diff
@@ -179,6 +179,9 @@ begin
 end
 
+lemma det_eq_zero_of_column_eq_zero {A : matrix n n R} (i : n) (h : ∀ j, A j i = 0) : det A = 0 :=
+by { rw ← det_transpose, exact det_eq_zero_of_row_eq_zero i h, }
```
```suggestion
lemma det_eq_zero_of_column_eq_zero {A : matrix n n R} (j : n) (h : ∀ i, A i j = 0) : det A = 0 :=
by { rw ← det_transpose, exact det_eq_zero_of_row_eq_zero j h, }
```

jcommelin reviewed (approved):
Thanks :tada:

bors d+

Commit:
Update src/linear_algebra/determinant.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
